lib/src/rbac/default_role_manager.dart

```diff
@@ -12,46 +12,114 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
+import 'domain_roles.dart';
 import 'role.dart';
 import 'role_manager.dart';
 
+const DEFAULT_DOMAIN = 'casbin::default';
+
 class DefaultRoleManager implements RoleManager {
-  Map<String, Role> allRoles;
+  Map<String, DomainRoles> allDomains;
   int maxHierarchyLevel;
+  bool hasPattern;
+  bool hasDomainPattern;
+  MatchingFunc? matchingFunc;
+  MatchingFunc? domainMatchingFunc;
 
   /// DefaultRoleManager is the constructor for creating an instance of the
   /// default RoleManager implementation.
   ///
   /// [maxHierarchyLevel] the maximized allowed RBAC hierarchy level.
-  DefaultRoleManager(this.maxHierarchyLevel) : allRoles = {};
+  DefaultRoleManager(
+    this.maxHierarchyLevel, [
+    this.matchingFunc,
+    this.domainMatchingFunc,
+  ])  : allDomains = {},
+        hasPattern = matchingFunc == null ? false : true,
+        hasDomainPattern = domainMatchingFunc == null ? false : true;
 
-  bool hasRole(String name) {
-    return allRoles.containsKey(name);
+  /// Clears all stored data and resets the role manager to the initial state.
+  @override
+  void clear() {
+    allDomains.clear();
+    allDomains[DEFAULT_DOMAIN] = DomainRoles();
   }
 
-  Role createRole(String name) {
-    return allRoles.putIfAbsent(name, () => Role(name));
+  String domainName([List<String> domain = const []]) {
+    return domain.isEmpty ? DEFAULT_DOMAIN : domain[0];
   }
 
-  /// Clears all stored data and resets the role manager to the initial state.
-  @override
-  void clear() {
-    allRoles.clear();
+  DomainRoles getOrCreateDomainRoles(final String domain) {
+    return allDomains.putIfAbsent(domain, () => DomainRoles());
+  }
+
+  void addMatchingFunc(dynamic name, MatchingFunc? func) {
+    hasPattern = true;
+
+    if (name.runtimeType == String || func != null) {
+      matchingFunc = func;
+    } else if (name.runtimeType == MatchingFunc) {
+      matchingFunc = name;
+    } else {
+      throw Exception('error: domain should be 1 parameter');
```

Comment:
I would prefer an `ArgumentError` here.

Reply:
Also, the error message could be clearer in this case.

Reply:
I've made all the required changes. Please verify.